core/src/main/java/io/opentelemetry/android/export/BufferedDelegatingExporter.kt

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.android.export
+
+import io.opentelemetry.sdk.common.CompletableResultCode
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * An in-memory buffer delegating signal exporter that buffers signal in memory until a delegate is set.
+ * Once a delegate is set, the buffered signals are exported to the delegate.
+ *
+ * The buffer size is set to 5,000 by default. If the buffer is full, the exporter will drop new signals.
+ */
+internal abstract class BufferedDelegatingExporter<T, D>(private val bufferedSignals: Int = 5_000) {
+    @Volatile
+    private var delegate: D? = null
+    private val buffer = arrayListOf<T>()
+    private val lock = Any()
```

Comment:
I'm a little concerned/confused about the concurrency handling here. I appreciate that effort was put into this, but I don't completely follow it. 

1. The delegate is both volatile and also sometimes (but not always) guarded by the `lock`. Do we really need both?  Can we perhaps simplify this by removing volatile and always use the `lock` when accessing the `delegate`?
2. The buffer that stores the telemetry doesn't seem to have any concurrency provisions...and I think maybe it should. Might be as simple as using a `ConcurrentLinkedQueue` or something?

Reply:
Sure, happy to elaborate.

1. Volatile is needed to correctly implement the double check pattern. You can read more about it [here](https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html).
2. It's not entirely obvious but the only time the block is called with a `null` delegate is when it's inside the `synchronized` block so no additional synchronization is required. This code was written with high performance in mind but this could perhaps be documented if nothing else.

Reply:
My understanding of the DCL pattern is that it's an anti-pattern. I agree with @breedx-splk that it's probably best to use synchronization only.

Reply:
The main reason this was chosen is that, as a library, it's important to think about performance, especially in low level functionality like this. Users could potentially be logging a lot and synchronizing would incur a cost for every log entry and albeit tiny, could add up.

I also wouldn't say it's an anti pattern. It's well established and used frequently (for example by [Dagger](https://github.com/google/dagger/blob/master/java/dagger/internal/DoubleCheck.java)) It's complicated and should be avoided unless needed, but I think it's warranted here.

Unless you want to also lock while dispatching to the delegate (which would be a risky idea) you'd end up with fairly similar code, just lacking the outer check and the volatile statement:
```kotlin
val delegate = synchronized(lock) {
  delegate ?: run {
    writeToBuffer(entry)
    return
  }
}
delegate.export(entry)
```

Reply:
That being said, this is your library and I'm happy to change this to whatever you think is best. I wanted to offer an explanation of why we chose this pattern but ultimately you're the ones who have to main this so the call is yours to make.

Reply:
Got it, thank you for the explanation. I agree that we should make an effort to ensure that this library doesn't have a poor performance and I appreciate the thoughtfulness put into this contribution in that regard! My concern is that, generally speaking, attempting to do premature optimizations might make the code a bit more difficult to understand and in the worst case might cause some other issues that are difficult to spot, especially for concurrency-related code as liveness issues might arise under specific conditions only, so that's why I think we should start simple and then gradually make things more complex as the need arises. In this case, if during your tests you've spotted some considerable performance-related issues that are only addressable by using this approach, then I think we can keep it, otherwise, I'd prefer to start with the option of using synchronization only.

Reply:
A bit late to the game, but I think it's reasonable to take extra care in handling thorny race conditions during transition. If the transition is unclean, there could be buffered data that gets lost, so locking this up, even with some more complicated code, is a fine thing to do. We might want to add more robust testing of edge cases where we simulate race conditions (good times), but I think it's beneficial overall to have something like this to increase the guarantee.

That said, are your concerns, @breedx-splk and @LikeTheSalad , more with readability or with the approach, specifically with the locking at every export call?

Another approach could be to lock only during the transition when the delegate is being set and not lock during export. When the transition is done, we can flush the in-memory data into the delegate, taking extra care on the in-memory queue to ensure that data coming in during the transition is picked picked. We just have to ensure that there isn't anything left in the in-memory queue after the transition, even if it comes in late.

What do you think, @ansman?

Reply:
I have a rewrite that doesn't use double checking and I think it's fine. I don't think there is any reduction in edge cases with the double locking, it would be a pure performance improvement.

One thing that I do want to ask is if it's ok to call the delegate while holding the lock. This is risky because if the delegate ends up waiting for something that in turn tries to log something you'd get a deadlock. This is true with or without the double locking I believe. 

But on the other hand, calling the delegate without holding the lock can lead to out of order exports during initialization. I don't think the latter is a big deal perhaps and preferable over deadlocking. But It's also possible that this isn't a scenario that can actually happen so let me know how you want to proceed.

Reply:
There's no guarantee of ordering in terms of export, nor the order in which it's received in the Collector, so as long as everything that's buffered gets exported without an unreasonable delay, I think that's all the guarantee we'd need. And it eliminates any possibility of a deadlock by not requiring the lock to be held :-)

Reply:
I agree with @bidetofevil, the order shouldn't matter in this case, the signal's timestamp should be used later when querying the data to ensure they're properly ordered. So I think the option that doesn't have the deadlock risk is the one we should go with.